hs_core/views/utils.py

```diff
@@ -559,84 +559,116 @@ def get_metadata_contenttypes():
-                                  then=Value(True, BooleanField()))))
+    if not filter or 'owned' in filter:
+        # get a list of resources with effective OWNER privilege
+        owned_resources = user.uaccess.get_resources_with_explicit_access(PrivilegeCodes.OWNER)
+
+    if not filter or 'editable' in filter:
+        # get a list of resources with effective CHANGE privilege (should include resources that the
+        # user has access to via group
+        editable_resources = user.uaccess.get_resources_with_explicit_access(PrivilegeCodes.CHANGE,
+                                                                             via_group=True)
+
+    if not filter or 'viewable' in filter:
+        # get a list of resources with effective VIEW privilege (should include resources that the
+        # user has access via group
+        viewable_resources = user.uaccess.get_resources_with_explicit_access(PrivilegeCodes.VIEW,
+                                                                             via_group=True)
 
+    if not filter or 'discovered' in filter:
+        discovered_resources = user.ulabels.my_resources
+        if annotate:
+            discovered_resources = discovered_resources.all().annotate(discovered=Value(True, BooleanField()))
+
+    favorite_resources = user.ulabels.favorited_resources
+
+    # join all queryset objects.
+    resource_collection = owned_resources.distinct() | \
+        editable_resources.distinct() | \
+        viewable_resources.distinct() | \
+        discovered_resources.distinct()
+    if not filter or 'favorites' in filter:
+        resource_collection = resource_collection | favorite_resources.distinct()
+
+    # remove obsoleted resources
+    resource_collection = resource_collection.exclude(object_id__in=Relation.objects.filter(
+            type='isReplacedBy').values('object_id')).exclude(extra_data__to_be_deleted__isnull=False)
+
+    # When used in the My Resources page, annotate for speed
+    if annotate:
         # The annotated field 'has_labels' would allow us to query the DB for labels only if the
         # resource has labels - that means we won't hit the DB for each resource listed on the page
         # to get the list of labels for a resource
+        labeled_resources = user.ulabels.labeled_resources
         resource_collection = resource_collection.annotate(has_labels=Case(
             When(short_id__in=labeled_resources.values_list('short_id', flat=True),
                  then=Value(True, BooleanField()))))
 
-        resource_collection = resource_collection.only('short_id', 'title', 'resource_type', 'created')
+        resource_collection = resource_collection.annotate(
+            is_favorite=Case(When(short_id__in=favorite_resources.values_list('short_id', flat=True),
+                                    then=Value(True, BooleanField()))))
+
+        resource_collection = resource_collection.only('short_id', 'title', 'resource_type', 'created', 'updated')
```

Comment:
I realized if we include the `'slug'` attribute here that should take care of not generating an extra query in the template (`{{res.get_absolute_url}}`). Sorry for my previous suggestion to fix this extra query problem.

Reply:
Good point, thanks for pointing this out (good for me to learn). I think we are okay as it is now, but if you prefer I can use `slug` instead